packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+      recipientKeys: [ed25519Key.publicKeyBase58],
+      serviceEndpoint,
+      accept: ['didcomm/aip2;env=rfc19'],
+      // It is important that we encode the routing keys as key references.
+      // So instead of using plain verkeys, we should encode them as did:key dids
+      routingKeys: mediatorRoutingKey ? [mediatorRoutingKey] : [],
+    })
+
+    const didDocument = new DidDocumentBuilder('')
+      .addAuthentication(ed25519VerificationMethod)
+      .addKeyAgreement(x25519VerificationMethod)
+      .addService(service)
+      .build()
+
+    const peerDid = DidPeer.fromDidDocument(didDocument, PeerDidNumAlgo.GenesisDoc)
+
+    const didRecord = new DidRecord({
+      id: peerDid.did,
+      role: DidDocumentRole.Created,
+      // It is important to take the did document from the PeerDid class
+      // as it will have the id property
+      // Should not we also resolve and store document for inline peer did?
+      didDocument: peerDid.numAlgo === PeerDidNumAlgo.GenesisDoc ? peerDid.didDocument : undefined,
+      tags: {
+        // We need to save the recipientKeys, so we can find the associated did
+        // of a key when we receive a message from another connection.
+        recipientKeys: peerDid.didDocument.recipientKeys,
+      },
+    })
+
+    await this.didRepository.save(didRecord)
+    this.logger.debug('Did record created.', didRecord)
+    return peerDid
+  }
+
+  private async createSignedAttachment(didDoc: DidDocument, verkey: string) {
+    const didDocAttach = new Attachment({
+      mimeType: 'application/json',
+      data: new AttachmentData({
+        base64: JsonEncoder.toBase64(didDoc),
+      }),
+    })
+
+    const kid = Key.fromPublicKeyBase58(verkey, KeyType.Ed25519)
+    const payload = JsonEncoder.toBuffer(didDoc)
+
+    const jws = await this.jwsService.createJws({
+      payload,
+      verkey,
+      header: {
+        kid,
+      },
+    })
+
+    didDocAttach.addJws(jws)
+    return didDocAttach
+  }
+
+  private async verifyAttachmentSignature(didDocAttachment: Attachment, invitationKeys: string[] = []) {
+    const jws = didDocAttachment.data.jws
+
+    if (!jws) {
+      throw new ProblemReportError('DidDoc signature is missing.', { problemCode: 'request_not_accepted' })
+    }
+
+    const json = didDocAttachment.getDataAsJson() as Record<string, unknown>
+    this.logger.trace('DidDocument JSON', json)
+
+    const payload = JsonEncoder.toBuffer(json)
+    const didDocument = JsonTransformer.fromJSON(json, DidDocument)
+
+    const { isValid, signerVerkeys } = await this.jwsService.verifyJws({ jws, payload })
+
+    const didDocKeys = didDocument.authentication
+      .map((a) => {
+        if (typeof a === 'string') {
+          return a
+        }
+        // TODO return key according to type property?
+        return a.publicKeyBase58
+      })
+      .concat(invitationKeys)
```

Comment:
As discusses during our call yesterday. If `a` is of type string it means it is a reference to an item in the `verificationMethod` array. We should dereference it.

But while we're at it, it would be good to use the generic way to get the base58Key:

```suggestion
    const didDocKeys = didDocument.authentication
      .map((authentication) => {
        const dereferenced = typeof authentication === 'string' ? didDocument.dereferenceKey(authentication) : authentication))
        const { getKeyFromVerificationMethod } = getKeyDidMappingByVerificationMethod(entry)
        const key = getKeyFromVerificationMethod(entry)
        
        return key.publicKeyBase58
      })
      .concat(invitationKeys)
```

Reply:
Yeah, thanks for that code snippet 👍  It would take me much longer to understand how to compose those functions to get the key. Updated.